### ISSUE TITLE
Provide a way to disable a function outside of function.json

### ISSUE THREAD

**Issue by lindydonna** [Azure/azure-webjobs-sdk-script#1331]
For the new C# programming model, we need a way to disable functions outside of function.json. The enable/disable state should be provided as part of the API: https://github.com/projectkudu/kudu/wiki/Functions-API 

There should also be the option of disabling all functions in the Function App. This is important for deployment slot scenarios. See https://github.com/Azure/azure-webjobs-sdk-script/issues/1729

As part of this, we should remove the code path that uses `[Disable]` attribute when `configurationSource` is set to `attributes`.

**Comment by vkhante** [comment]
We will need to disable the function in ARM template for slicing work we want to do.

**Comment by christopheranderson** [comment]
Need more user experience guidelines on this. How would the customer disable it, etc?

**Comment by lindydonna** [comment]
Adding internal discussion notes:

@davidebbo says:

Thanks Donna. It’s worth pointing out that this will be useful beyond the precompiled scenario. Right now, even with a plain Node function, you blow away your ‘disabled’ state each time you redeploy.

For the options, #1 and #2 are indeed not good choices (for the reasons you give).

One tricky thing we need to think about is how this will play with slots. E.g.
-	I disable a function in staging
-	I swap

Should that function now be disabled in Staging or in Prod? With what we have today, the Disabled state always moves with the swap, but let’s think about whether that’s always what the user wants (case could be made for either).

If we’re going to store that outside of wwwroot, the obvious place is the data folder. E.g. D:\home\data\Functions\settings, just like we have D:\home\data\Functions\secrets (with one json file per function).

--- 

@paulbatum says:

Yes data gets swapped too. 

I think ideally, the disabled behavior is slot specific and sticky. I can imagine having my staging slot setup such that it points to production resources, but the functions are disabled and I run them manually when I'm doing last minute verification. 

I'm not sure what a good solution for this would look like. Neither the file system nor app settings work well.

If we can't get this behavior, then we're going to need to rethink our portal UX around disabling. Because right now we list the functions and their disabled state as a child of the app/slot. You would not expect a change in one to have an effect on the other.

-----

@davidebbo says:

Exactly, the need for stickiness is what makes it tricky.

Interestingly, this is very similar to what happened with keys:
-	We had them in the data folder
-	As we added slots, we realized we wanted stickiness, which is painful to achieve with files
-	So we moved them to storage (currently that only happens if you opt into slots)

Does that mean that this ‘disabled’ state should also live in storage? Logically, it should. But in practice I’m not a fan of that, as we really should be reducing our storage reliance, and not adding new dependencies on it. E.g. we know we can’t do this in the Logic App xslt scenario, where we can’t have a connection string.

So where does that leave us? It feels like we need a new mechanism that has the following properties:
-	Storage is in the Antares DB, and is automatically replicated across stamps (just like all the other site properties like app settings, .net version, hostnames…)
-	Behaves in a sticky way (e.g. like hostnames)
-	The state is reflected in generated TEMP files (much like applicationhost.config)
-	Changes get propagated by DWAS with no App Restart
-	There is a way to update those from inside the sandbox, via a FE based mechanism similar to what we do with AutoSwap and Function Sync

Frankly, if we had that, we would use it for keys as well instead of storage.

But that’s a fairly sizeable investment to do E2E.

----

@davidebbo says:

After some internal chatting, one key conclusion is that we don’t dislike storage in general, but we dislike having to rely on a full storage connection string (which kills the Logic App-like scenarios).

But if we can instead rely on sas URL, we start having options. Today, we use a full storage connection string for the keys. However, we think it should work via a sas (unless Azure Files, which definitely doesn’t support sas). So we could use storage for the ‘disabled’ state in a similar way to what we do for keys, if we make sure that it works E2E with sas. In fact, it would make sense to have a more general settings abstraction to deal with any such situation going forward.
